app/src/test/java/me/hufman/androidautoidrive/music/SpotifyAuthStateManagerTest.kt

```diff
@@ -0,0 +1,250 @@
+
+		assertEquals(authState, spotifyAuthStateManager.currentState)
+	}
+
+	@Test
+	fun testNewInstance_DeserializeJsonException() {
+		val state = "authStateJson"
+		appSettings[AppSettings.KEYS.SPOTIFY_AUTH_STATE_JSON] = state
+
+		PowerMockito.mockStatic(AuthState::class.java)
+		PowerMockito.`when`(AuthState.jsonDeserialize(state)).thenAnswer { throw JSONException("message") }
+
+		spotifyAuthStateManager = Whitebox.invokeConstructor(SpotifyAuthStateManager::class.java, appSettings)
+
+		assertTrue(isEmptyAuthState(spotifyAuthStateManager.currentState))
+	}
+
+	@Test
+	fun testGetAuthorizationCode_NullLastAuthorizationResponse() {
+		val currentState: AuthState = mock()
+		whenever(currentState.lastAuthorizationResponse).thenReturn(null)
+		FieldSetter.setField(spotifyAuthStateManager, SpotifyAuthStateManager::class.java.getDeclaredField("currentState"), currentState)
+
+		assertNull(spotifyAuthStateManager.getAuthorizationCode())
+	}
+
+	@Test
+	fun testGetAuthorizationCode_ExistingLastAuthorizationResponse() {
+		val authorizationCode = "authorizationCode"
+		val authorizationResponse: AuthorizationResponse = mock()
+		FieldSetter.setField(authorizationResponse, AuthorizationResponse::class.java.getDeclaredField("authorizationCode"), authorizationCode)
+
+		val currentState: AuthState = mock()
+		whenever(currentState.lastAuthorizationResponse).thenReturn(authorizationResponse)
+		FieldSetter.setField(spotifyAuthStateManager, SpotifyAuthStateManager::class.java.getDeclaredField("currentState"), currentState)
+
+		assertEquals(authorizationCode, spotifyAuthStateManager.getAuthorizationCode())
+	}
+
+	@Test
+	fun testGetAccessToken() {
+		val accessToken = "accessToken"
+		val currentState: AuthState = mock()
+		whenever(currentState.accessToken).thenReturn(accessToken)
+		FieldSetter.setField(spotifyAuthStateManager, SpotifyAuthStateManager::class.java.getDeclaredField("currentState"), currentState)
+
+		assertEquals(accessToken, spotifyAuthStateManager.getAccessToken())
+	}
+
+	@Test
+	fun testGetAccessTokenExpirationIn_NullAccessTokenExpirationTime() {
+		val currentState: AuthState = mock()
+		whenever(currentState.accessTokenExpirationTime).thenReturn(null)
+		FieldSetter.setField(spotifyAuthStateManager, SpotifyAuthStateManager::class.java.getDeclaredField("currentState"), currentState)
+
+		assertEquals(0, spotifyAuthStateManager.getAccessTokenExpirationIn())
+	}
+
+	@Test
+	fun testGetAccessTokenExpirationIn_ExistingAccessTokenExpirationTime() {
+		val expirationTime = 5000L
+		val currentTime = 3000L
+		val currentState: AuthState = mock()
+		whenever(currentState.accessTokenExpirationTime).thenReturn(expirationTime)
+		FieldSetter.setField(spotifyAuthStateManager, SpotifyAuthStateManager::class.java.getDeclaredField("currentState"), currentState)
+
+		PowerMockito.mockStatic(System::class.java)
+		PowerMockito.`when`(System.currentTimeMillis()).doAnswer { currentTime }
```

Comment:
Oooo this is powerful

Reply:
Yeah PowerMockito is really amazing, hopefully it will help us out with testing and allow us to hit higher code coverage for things that weren't as easy to test before. It also makes singletons very easy to test which is great!